{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocessing of the Log File\n",
    "#### We will preprocessing and wrangle the messy log files and cleanly populate the dataframe. This dataframe will be used to visualise our data in our web app using Dash Plotly framework."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from datetime import datetime"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read contents of the file to the program"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n",
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n",
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 3343, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-56-1cd3dc4ebeaf>\", line 1, in <module>\n",
      "    file = open(\"gerror.txt\", \"r\")\n",
      "FileNotFoundError: [Errno 2] No such file or directory: 'gerror.txt'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2044, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'FileNotFoundError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 1169, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 316, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 350, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/usr/lib/python3.6/inspect.py\", line 1490, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"/usr/lib/python3.6/inspect.py\", line 1448, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/usr/lib/python3.6/inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"/usr/lib/python3.6/inspect.py\", line 725, in getmodule\n",
      "    file = getabsfile(object, _filename)\n",
      "  File \"/usr/lib/python3.6/inspect.py\", line 709, in getabsfile\n",
      "    return os.path.normcase(os.path.abspath(_filename))\n",
      "  File \"/usr/lib/python3.6/posixpath.py\", line 383, in abspath\n",
      "    cwd = os.getcwd()\n",
      "FileNotFoundError: [Errno 2] No such file or directory\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 3343, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-56-1cd3dc4ebeaf>\", line 1, in <module>\n",
      "    file = open(\"gerror.txt\", \"r\")\n",
      "FileNotFoundError: [Errno 2] No such file or directory: 'gerror.txt'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2044, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'FileNotFoundError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 3263, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 3360, in run_code\n",
      "    self.showtraceback(running_compiled_code=True)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2047, in showtraceback\n",
      "    value, tb, tb_offset=tb_offset)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 1436, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 1336, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 1193, in structured_traceback\n",
      "    tb_offset)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 1150, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 451, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2044, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 1169, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 316, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 350, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/usr/lib/python3.6/inspect.py\", line 1490, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"/usr/lib/python3.6/inspect.py\", line 1448, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/usr/lib/python3.6/inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"/usr/lib/python3.6/inspect.py\", line 725, in getmodule\n",
      "    file = getabsfile(object, _filename)\n",
      "  File \"/usr/lib/python3.6/inspect.py\", line 709, in getabsfile\n",
      "    return os.path.normcase(os.path.abspath(_filename))\n",
      "  File \"/usr/lib/python3.6/posixpath.py\", line 383, in abspath\n",
      "    cwd = os.getcwd()\n",
      "FileNotFoundError: [Errno 2] No such file or directory\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 3343, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-56-1cd3dc4ebeaf>\", line 1, in <module>\n",
      "    file = open(\"gerror.txt\", \"r\")\n",
      "FileNotFoundError: [Errno 2] No such file or directory: 'gerror.txt'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2044, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'FileNotFoundError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 3263, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 3360, in run_code\n",
      "    self.showtraceback(running_compiled_code=True)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2047, in showtraceback\n",
      "    value, tb, tb_offset=tb_offset)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 1436, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 1336, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 1193, in structured_traceback\n",
      "    tb_offset)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 1150, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 451, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2044, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2895, in _run_cell\n",
      "    return runner(coro)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/async_helpers.py\", line 68, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 3072, in run_cell_async\n",
      "    interactivity=interactivity, compiler=compiler, result=result)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 3282, in run_ast_nodes\n",
      "    self.showtraceback()\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2047, in showtraceback\n",
      "    value, tb, tb_offset=tb_offset)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 1436, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 1336, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 1211, in structured_traceback\n",
      "    chained_exceptions_tb_offset)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 1150, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 451, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py\", line 2044, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 1169, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 316, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/home/pako/.local/lib/python3.6/site-packages/IPython/core/ultratb.py\", line 350, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/usr/lib/python3.6/inspect.py\", line 1490, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"/usr/lib/python3.6/inspect.py\", line 1448, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/usr/lib/python3.6/inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"/usr/lib/python3.6/inspect.py\", line 725, in getmodule\n",
      "    file = getabsfile(object, _filename)\n",
      "  File \"/usr/lib/python3.6/inspect.py\", line 709, in getabsfile\n",
      "    return os.path.normcase(os.path.abspath(_filename))\n",
      "  File \"/usr/lib/python3.6/posixpath.py\", line 383, in abspath\n",
      "    cwd = os.getcwd()\n",
      "FileNotFoundError: [Errno 2] No such file or directory\n"
     ]
    }
   ],
   "source": [
    "file = open(\"gerror.txt\", \"r\")\n",
    "contents = \"\"\n",
    "for i in range(5000):\n",
    "    contents += str(file.readline())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data wrangling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERROR! Session/line number was not unique in database. History logging moved to new session 35\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>date_time</th>\n",
       "      <th>day_of_the_week</th>\n",
       "      <th>error_messages</th>\n",
       "      <th>severity_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2019-05-02</td>\n",
       "      <td>Thu</td>\n",
       "      <td>user authentication failure for password misma...</td>\n",
       "      <td>[auth_basic:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2019-05-27</td>\n",
       "      <td>Mon</td>\n",
       "      <td>php  undefined salt in on line</td>\n",
       "      <td>[:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2019-05-27</td>\n",
       "      <td>Mon</td>\n",
       "      <td>php  undefined salt in on line</td>\n",
       "      <td>[:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2019-06-07</td>\n",
       "      <td>Fri</td>\n",
       "      <td>php  undefined salt in on line</td>\n",
       "      <td>[:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2019-06-07</td>\n",
       "      <td>Fri</td>\n",
       "      <td>php  undefined salt in on line</td>\n",
       "      <td>[:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2019-06-07</td>\n",
       "      <td>Fri</td>\n",
       "      <td>php  undefined salt in on line</td>\n",
       "      <td>[:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2020-02-29</td>\n",
       "      <td>Sat</td>\n",
       "      <td>user authentication failure for password misma...</td>\n",
       "      <td>[auth_basic:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2020-03-01</td>\n",
       "      <td>Sun</td>\n",
       "      <td>user authentication failure for password</td>\n",
       "      <td>[auth_basic:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2020-03-01</td>\n",
       "      <td>Sun</td>\n",
       "      <td>user authentication failure for password</td>\n",
       "      <td>[auth_basic:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2020-03-01</td>\n",
       "      <td>Sun</td>\n",
       "      <td>user authentication failure for password</td>\n",
       "      <td>[auth_basic:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2020-03-01</td>\n",
       "      <td>Sun</td>\n",
       "      <td>user authentication failure for password</td>\n",
       "      <td>[auth_basic:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2020-03-01</td>\n",
       "      <td>Sun</td>\n",
       "      <td>user authentication failure for password</td>\n",
       "      <td>[auth_basic:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2020-03-01</td>\n",
       "      <td>Sun</td>\n",
       "      <td>user authentication failure for password</td>\n",
       "      <td>[auth_basic:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2020-03-01</td>\n",
       "      <td>Sun</td>\n",
       "      <td>user authentication failure for password</td>\n",
       "      <td>[auth_basic:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2020-03-01</td>\n",
       "      <td>Sun</td>\n",
       "      <td>user authentication failure for password</td>\n",
       "      <td>[auth_basic:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2020-03-01</td>\n",
       "      <td>Sun</td>\n",
       "      <td>user authentication failure for password</td>\n",
       "      <td>[auth_basic:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>2020-03-01</td>\n",
       "      <td>Sun</td>\n",
       "      <td>user authentication failure for password</td>\n",
       "      <td>[auth_basic:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>2020-03-01</td>\n",
       "      <td>Sun</td>\n",
       "      <td>user authentication failure for password</td>\n",
       "      <td>[auth_basic:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>2020-03-01</td>\n",
       "      <td>Sun</td>\n",
       "      <td>user authentication failure for password</td>\n",
       "      <td>[auth_basic:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>2020-03-01</td>\n",
       "      <td>Sun</td>\n",
       "      <td>user authentication failure for password</td>\n",
       "      <td>[auth_basic:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>2020-03-01</td>\n",
       "      <td>Sun</td>\n",
       "      <td>user authentication failure for password</td>\n",
       "      <td>[auth_basic:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>2020-03-01</td>\n",
       "      <td>Sun</td>\n",
       "      <td>user authentication failure for password</td>\n",
       "      <td>[auth_basic:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>2020-03-01</td>\n",
       "      <td>Sun</td>\n",
       "      <td>user authentication failure for password</td>\n",
       "      <td>[auth_basic:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>2020-03-09</td>\n",
       "      <td>Mon</td>\n",
       "      <td>script not found or unable to stat</td>\n",
       "      <td>[:error]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker shared already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker local already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker shared already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker local already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker shared already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker local already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4970</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker shared already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4971</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker local already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4972</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker shared already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4973</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker local already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4974</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker shared already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4975</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker local already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4976</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker shared already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4977</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker local already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4978</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker shared already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4979</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker local already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4980</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker shared already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4981</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker local already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4982</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker shared already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4983</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker local already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4984</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker shared already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4985</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker local already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4986</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker shared already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4987</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker local already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4988</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker shared already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4989</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker local already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4990</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker shared already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4991</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker local already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4992</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker shared already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4993</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker local already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4994</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker shared already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4995</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker local already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4996</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker shared already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4997</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker local already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4998</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker shared already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4999</th>\n",
       "      <td>2020-03-29</td>\n",
       "      <td>Sun</td>\n",
       "      <td>worker local already initialized</td>\n",
       "      <td>[proxy:debug]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5000 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       date_time day_of_the_week  \\\n",
       "0     2019-05-02             Thu   \n",
       "1     2019-05-27             Mon   \n",
       "2     2019-05-27             Mon   \n",
       "3     2019-06-07             Fri   \n",
       "4     2019-06-07             Fri   \n",
       "5     2019-06-07             Fri   \n",
       "6     2020-02-29             Sat   \n",
       "7     2020-03-01             Sun   \n",
       "8     2020-03-01             Sun   \n",
       "9     2020-03-01             Sun   \n",
       "10    2020-03-01             Sun   \n",
       "11    2020-03-01             Sun   \n",
       "12    2020-03-01             Sun   \n",
       "13    2020-03-01             Sun   \n",
       "14    2020-03-01             Sun   \n",
       "15    2020-03-01             Sun   \n",
       "16    2020-03-01             Sun   \n",
       "17    2020-03-01             Sun   \n",
       "18    2020-03-01             Sun   \n",
       "19    2020-03-01             Sun   \n",
       "20    2020-03-01             Sun   \n",
       "21    2020-03-01             Sun   \n",
       "22    2020-03-01             Sun   \n",
       "23    2020-03-09             Mon   \n",
       "24    2020-03-29             Sun   \n",
       "25    2020-03-29             Sun   \n",
       "26    2020-03-29             Sun   \n",
       "27    2020-03-29             Sun   \n",
       "28    2020-03-29             Sun   \n",
       "29    2020-03-29             Sun   \n",
       "...          ...             ...   \n",
       "4970  2020-03-29             Sun   \n",
       "4971  2020-03-29             Sun   \n",
       "4972  2020-03-29             Sun   \n",
       "4973  2020-03-29             Sun   \n",
       "4974  2020-03-29             Sun   \n",
       "4975  2020-03-29             Sun   \n",
       "4976  2020-03-29             Sun   \n",
       "4977  2020-03-29             Sun   \n",
       "4978  2020-03-29             Sun   \n",
       "4979  2020-03-29             Sun   \n",
       "4980  2020-03-29             Sun   \n",
       "4981  2020-03-29             Sun   \n",
       "4982  2020-03-29             Sun   \n",
       "4983  2020-03-29             Sun   \n",
       "4984  2020-03-29             Sun   \n",
       "4985  2020-03-29             Sun   \n",
       "4986  2020-03-29             Sun   \n",
       "4987  2020-03-29             Sun   \n",
       "4988  2020-03-29             Sun   \n",
       "4989  2020-03-29             Sun   \n",
       "4990  2020-03-29             Sun   \n",
       "4991  2020-03-29             Sun   \n",
       "4992  2020-03-29             Sun   \n",
       "4993  2020-03-29             Sun   \n",
       "4994  2020-03-29             Sun   \n",
       "4995  2020-03-29             Sun   \n",
       "4996  2020-03-29             Sun   \n",
       "4997  2020-03-29             Sun   \n",
       "4998  2020-03-29             Sun   \n",
       "4999  2020-03-29             Sun   \n",
       "\n",
       "                                         error_messages      severity_level  \n",
       "0     user authentication failure for password misma...  [auth_basic:error]  \n",
       "1                       php  undefined salt in on line             [:error]  \n",
       "2                       php  undefined salt in on line             [:error]  \n",
       "3                       php  undefined salt in on line             [:error]  \n",
       "4                       php  undefined salt in on line             [:error]  \n",
       "5                       php  undefined salt in on line             [:error]  \n",
       "6     user authentication failure for password misma...  [auth_basic:error]  \n",
       "7             user authentication failure for password   [auth_basic:error]  \n",
       "8             user authentication failure for password   [auth_basic:error]  \n",
       "9             user authentication failure for password   [auth_basic:error]  \n",
       "10            user authentication failure for password   [auth_basic:error]  \n",
       "11            user authentication failure for password   [auth_basic:error]  \n",
       "12            user authentication failure for password   [auth_basic:error]  \n",
       "13            user authentication failure for password   [auth_basic:error]  \n",
       "14            user authentication failure for password   [auth_basic:error]  \n",
       "15            user authentication failure for password   [auth_basic:error]  \n",
       "16            user authentication failure for password   [auth_basic:error]  \n",
       "17            user authentication failure for password   [auth_basic:error]  \n",
       "18            user authentication failure for password   [auth_basic:error]  \n",
       "19            user authentication failure for password   [auth_basic:error]  \n",
       "20            user authentication failure for password   [auth_basic:error]  \n",
       "21            user authentication failure for password   [auth_basic:error]  \n",
       "22            user authentication failure for password   [auth_basic:error]  \n",
       "23                  script not found or unable to stat             [:error]  \n",
       "24                   worker shared already initialized        [proxy:debug]  \n",
       "25                    worker local already initialized        [proxy:debug]  \n",
       "26                   worker shared already initialized        [proxy:debug]  \n",
       "27                    worker local already initialized        [proxy:debug]  \n",
       "28                   worker shared already initialized        [proxy:debug]  \n",
       "29                    worker local already initialized        [proxy:debug]  \n",
       "...                                                 ...                 ...  \n",
       "4970                 worker shared already initialized        [proxy:debug]  \n",
       "4971                  worker local already initialized        [proxy:debug]  \n",
       "4972                 worker shared already initialized        [proxy:debug]  \n",
       "4973                  worker local already initialized        [proxy:debug]  \n",
       "4974                 worker shared already initialized        [proxy:debug]  \n",
       "4975                  worker local already initialized        [proxy:debug]  \n",
       "4976                 worker shared already initialized        [proxy:debug]  \n",
       "4977                  worker local already initialized        [proxy:debug]  \n",
       "4978                 worker shared already initialized        [proxy:debug]  \n",
       "4979                  worker local already initialized        [proxy:debug]  \n",
       "4980                 worker shared already initialized        [proxy:debug]  \n",
       "4981                  worker local already initialized        [proxy:debug]  \n",
       "4982                 worker shared already initialized        [proxy:debug]  \n",
       "4983                  worker local already initialized        [proxy:debug]  \n",
       "4984                 worker shared already initialized        [proxy:debug]  \n",
       "4985                  worker local already initialized        [proxy:debug]  \n",
       "4986                 worker shared already initialized        [proxy:debug]  \n",
       "4987                  worker local already initialized        [proxy:debug]  \n",
       "4988                 worker shared already initialized        [proxy:debug]  \n",
       "4989                  worker local already initialized        [proxy:debug]  \n",
       "4990                 worker shared already initialized        [proxy:debug]  \n",
       "4991                  worker local already initialized        [proxy:debug]  \n",
       "4992                 worker shared already initialized        [proxy:debug]  \n",
       "4993                  worker local already initialized        [proxy:debug]  \n",
       "4994                 worker shared already initialized        [proxy:debug]  \n",
       "4995                  worker local already initialized        [proxy:debug]  \n",
       "4996                 worker shared already initialized        [proxy:debug]  \n",
       "4997                  worker local already initialized        [proxy:debug]  \n",
       "4998                 worker shared already initialized        [proxy:debug]  \n",
       "4999                  worker local already initialized        [proxy:debug]  \n",
       "\n",
       "[5000 rows x 4 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# split string respect to newline \n",
    "lines = contents.split('\\n') # list of log entries\n",
    "\n",
    "date_time_list = []\n",
    "severity_level_list = []\n",
    "dirty_list = []\n",
    "\n",
    "for line in lines:\n",
    "    pieces = line.split(\" \")\n",
    "    if len(pieces) > 11: # some lines are less than the minimum required length\n",
    "        date_time_list.append(\" \".join([pieces[dt] for dt in range(5)]))\n",
    "        severity_level_list.append(pieces[5])\n",
    "        dirty_list.append(\" \".join([pieces[e] for e in range(6,len(pieces),1)]))\n",
    "\n",
    "# Clean the actual error messages\n",
    "error_msg_list = []\n",
    "error_msg = \"\"\n",
    "for dirty in dirty_list:\n",
    "    dirty_pieces = dirty.split(\" \")\n",
    "    error_msg = \"\"\n",
    "    for dirt_piece in dirty_pieces:\n",
    "        if not any(char.isdigit() for char in dirt_piece):\n",
    "            if not any(not char.isalnum() for char in dirt_piece):\n",
    "                error_msg += dirt_piece + \" \"\n",
    "    if error_msg:\n",
    "        error_msg_list.append(error_msg)\n",
    "\n",
    "# Formate the date\n",
    "MONTH_MAP = {\n",
    "    # map words to digits\n",
    "    \"Jan\":\"01\",\"Feb\":\"02\",\"Mar\":\"03\",\"Apr\":\"04\",\"May\":\"05\",\"Jun\":\"06\",\n",
    "    \"Jul\":\"07\",\"Aug\":\"08\",\"Sep\":\"09\",\"Oct\":\"10\",\"Nov\":\"11\",\"Dec\":\"12\"\n",
    "}\n",
    "\n",
    "day_of_week_list = []\n",
    "date_list = []\n",
    "for date_time_str in date_time_list:\n",
    "    date_time_str = date_time_str.replace('[','')\n",
    "    date_time_str = date_time_str.replace(']','')\n",
    "    date_time_pieces = date_time_str.split(\" \")\n",
    "    day_of_week_list.append(date_time_pieces[0])\n",
    "    formatted_date_str = date_time_pieces[2] +\"/\"+ MONTH_MAP[date_time_pieces[1][:3]] +\"/\"+ date_time_pieces[4][2:]\n",
    "    # convert date string to actual date object and append to date list\n",
    "    date_list.append(datetime.strptime(formatted_date_str, '%d/%m/%y').date())\n",
    "          \n",
    "data = {\n",
    "    \"date_time\": date_list,\n",
    "    \"day_of_the_week\": day_of_week_list,\n",
    "    \"severity_level\": severity_level_list,\n",
    "    \"error_messages\": error_msg_list\n",
    "}\n",
    "\n",
    "df = pd.DataFrame(data)\n",
    "df['error_messages'] = df['error_messages'].apply(lambda msg: msg.lower()) \n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.6.9 64-bit",
   "language": "python",
   "name": "python36964bit3c684b9b280e460683b25251b1d816f1"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
